### PR TITLE
[opt] Cache loop-invariant global vars to local vars

### LINKED ISSUE #5350: [perf] Enable load store forwarding across loop hierarchy for vulkan performance.
The `store_to_load_forwarding` transformation in the `cfg_optimization` pass ignores ExternalPtrStmt. It causes performance issue in [the SPH case](https://github.com/taichi-dev/taichi/issues/5046). The CUDA backend is insensitive to this optimization, but Vulkan backend cannot handles this well even on the same hardware.

I think we can get a ~2.2x speed up for the SPH case when this optimization is enabled.

The related source code section is [here](https://github.com/taichi-dev/taichi/blob/master/taichi/ir/control_flow_graph.cpp#L264-L294).

### PR DESCRIPTION
**Pull request by lin-hitonami**
Related issue = fixes #5350 

Global variables can't be store-to-load forwarded after `lower-access` pass, so we need to do `simplify` before it. It should speed up the program in all circumstances.

Caching loop-invariant global vars to local vars sometimes speeds up the program yet some time lets the program run slower so I let it controlled by the compiler config.

FPS of Yu's program on RTX3080 on Vulkan:
Original: 19fps
Simplified before lower access: 30fps
Cached loop-invariant global vars to local vars: 41fps

**This PR does things as follows:**
1. Extract a base class `LoopInvariantDetector` from `LoopInvariantCodeMotion`. This class maintains information to detect whether a statement is loop-invariant.
2. Let LICM move `GlobalPtrStmt`, `ArgLoadStmt` and `ExternalPtrStmt` out of the loop so that they become loop-invariant.
3. Add `CacheLoopInvariantGlobalVars` to move out loop-invariant global variables that are loop-unique in the offloaded task.
4. Add pass `cache_loop_invariant_global_vars` after `demote_atomics` before `demote_dense_struct_fors` (because loop-uniqueness can't be correctly detected after `demote_dense_struct_fors`) and add a compiler config flag to control it.
5. Add pass `full_simplify` before `lower_access` to enable store-to-load forwarding for GlobalPtrs.

### CODE REVIEW

**Comment by bobcao3** [comment]
By doing this, we are essentially trading memory access latency with register pressure. It might not always be better especially when global variables are likely uniform, and the GPU will do uniform value optimizations. On most recent GPUs the latency difference to L1/L0 is not that high anyways... We might need heuristics and cost functions with this.

**Comment by lin-hitonami** [comment]
> By doing this, we are essentially trading memory access latency with register pressure. It might not always be better especially when global variables are likely uniform, and the GPU will do uniform value optimizations. On most recent GPUs the latency difference to L1/L0 is not that high anyways... We might need heuristics and cost functions with this.

For @YuCrazing 's case in https://github.com/taichi-dev/taichi/issues/5046, the original version runs ~19fps on my machine, after fixing load-to-store forwarding of global ptrs, it runs ~29fps, then after caching the global vars to local vars, it runs~38fps. So at least it is useful for this case. We may need to test more examples.

**Comment by lin-hitonami** [comment]
> By doing this, we are essentially trading memory access latency with register pressure. It might not always be better especially when global variables are likely uniform, and the GPU will do uniform value optimizations. On most recent GPUs the latency difference to L1/L0 is not that high anyways... We might need heuristics and cost functions with this.

For @turbo0628 's case in https://github.com/taichi-dev/taichi/issues/5350#issuecomment-1198839839, it indeed becomes slower in my machine... I'm not very familiar with heuristics so I need some time to learn it. Maybe we can make it an option in the compile config first...

**Comment by turbo0628** [comment]
> By doing this, we are essentially trading memory access latency with register pressure. It might not always be better especially when global variables are likely uniform, and the GPU will do uniform value optimizations. On most recent GPUs the latency difference to L1/L0 is not that high anyways... We might need heuristics and cost functions with this.

It's pretty tricky that the performance gap  (23 vs 41 fps) is far more significant than L1/L0 difference. It behaves as if L1 cannot properly handle the accesses, and that's why we need this optimization at least for SPH kernels. 

Also, CUDA backend doesn't have this problem, though the PTX code still uses a lot of `ld`.

Could it be an L1 problem for Vulkan backend?

**Comment by bobcao3** [comment]
> 

Could be the on-GPU code optimization did this for the PTX

**Comment by lin-hitonami** [comment]
> > By doing this, we are essentially trading memory access latency with register pressure. It might not always be better especially when global variables are likely uniform, and the GPU will do uniform value optimizations. On most recent GPUs the latency difference to L1/L0 is not that high anyways... We might need heuristics and cost functions with this.
> 
> For @turbo0628 's case in [#5350 (comment)](https://github.com/taichi-dev/taichi/issues/5350#issuecomment-1198839839), it indeed becomes slower in my machine... I'm not very familiar with heuristics so I need some time to learn it. Maybe we can make it an option in the compile config first...

Now after https://github.com/taichi-dev/taichi/pull/6136 and https://github.com/taichi-dev/taichi/pull/6129 the regression has been fixed. 

Current FPS of Yu's SPH program on RTX3080 on Vulkan:
Original: 25fps
Simplified before lower access: 41fps
Cached loop-invariant global vars to local vars: 47fps

So should we let this pass default on?

File: taichi/transforms/loop_invariant_detector.h [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      return false;
+    }
+    if (current_scope != loop_blocks.top()) {
+      // If we enable moving code from a nested if block, we need to check
+      // visibility. Example:
+      // for i in range(10):
+      //   a = x[0]
+      //   if b:
+      //     c = a + 1
+      // Since we are moving statements outside the cloest for scope,
```
Comment (ailzhang) [inline taichi/transforms/loop_invariant_detector.h:36]:
```suggestion
      // Since we are moving statements outside the closest for scope,
```

**Review by ailzhang — APPROVED** [review]
Thanks!

File: taichi/transforms/cache_loop_invariant_global_vars.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
+#include "taichi/transforms/loop_invariant_detector.h"
+#include "taichi/ir/analysis.h"
+
+TLANG_NAMESPACE_BEGIN
```
[the remark this answers is not in the document]
Comment (lin-hitonami) [reply taichi/transforms/cache_loop_invariant_global_vars.cpp:4]: emmm... then we need to add `using namespace taichi::lang` or add `taichi::lang::` to everything defined in other files to make the code work...

[the remark this answers is not in the document]
Comment (ailzhang) [reply taichi/transforms/cache_loop_invariant_global_vars.cpp:4]: just add explicit namespaces ;) ?
```
namespace taichi {
namespace lang{

}
}
```

**Pull request merged**